It Came From Outer Spess: Adds midround changelings, delivered by an absolutely disgusting changeling meteor

Rhials opened the pull request:
## About The Pull Request

Adds a new dynamic midround opportunity and random event - Space Changeling. 

https://user-images.githubusercontent.com/28870487/215284465-f5c5c1b1-b83d-471a-89be-1b65a4d2f2d4.mp4

If you are fortunate enough to recieve this role, you will be stuffed into a changeling meteor and hurled at the side of the station. With no crew identities, no access, and no equipment, you'll have to rely on your **free** organic space suit and armblade to infiltrate the station and get settled.

With no disguises to fall back on, the midround changeling experience may lead to some very unfavorable situations. It's not unlikely that you'll be spotted making your way inside, or that someone will see the impact site and cause a panic. This role is not easy, but keep in mind that you also have nothing to lose in the event that you use Lesser Form/Headslug.

Aside from the starting circumstances, you have the same objectives and capabilities as a roundstart changeling. Getting inside of the station will be the hard part, but from there you can do what changelings do best and blend in.

<details>
<summary>A brief note on the free stuff you get:</summary>
<br>
You get the organic space suit and armblade for free. The space suit is absolutely vital, but I decided that the armblade should be given for free as well. It's necessary for breaking open windows or airlocks and getting access to the station, since otherwise your options are limited to arrivals/departures. Having to pay a 2 point tax to avoid walking naked into the main hallways of the station and getting gibbed is lame, and with the added difficulty of the role I think it's fair.
</details>

Also, this is my 100th PR here! :)

## Why It's Good For The Game

Adds midround changelings in a WAY COOLER way than just making a random crew/new arrival a changeling. 

Lets people experience Hardmode Changeling, and test the adaptability and flexibility of the most versatile antagonist even harder than before. Losing the option to bypass the whole shape-shifter thing by disguising as your crew identity presents a welcome change to the formula.

Adds a teensy bit more midround variety, so we stop getting Nightmare At The Thirty Minute Mark every round.
## Changelog
:cl: Rhials
add: Midround changeling spawn event.
add: Changeling meteor. It has a present for you.
/:cl:

Rhials committed:
big pr wow

Rhials committed:
oooooooops runtimes

Rhials committed:
it works now wow

you get instantly destroyed when the meteor deletes  but lol

Rhials committed:
it works even more now!

Rhials committed:
wake up wake up

Rhials committed:
makes the ling meteor a meateor subtype

Rhials committed:
adds meateor sprite wow

Rhials committed:
changeling now spawns with its suit equipped

Rhials committed:
subtype oopsies

Rhials committed:
when the imposter is sus

Rhials committed:
adds space changeling ruleset

puts the changeling meteor generation stuff into a global proc

Rhials committed:
pathing oopsioes

Rhials committed:
text change

Rhials committed:
tgui progress??

Rhials committed:
midround movement

Rhials committed:
work on this later

Rhials committed:
oops

Rhials committed:
readds the space suit outfit and auto-equips it

Rhials committed:
better spawn logic (works on runtimestation now!)

Rhials committed:
Merge remote-tracking branch 'upstream/master' into latejoin-ling

Rhials committed:
structure is modelled after the spider midround

adds free purchase option to purchase_power

makes code better

Rhials committed:
yeah

Rhials committed:
even better spawn logic, better comments

fixxes the midround pref thing

Rhials committed:
midround pref FINALLY works now

the fucking meteor MISSED RUNTIMESTATION TWICE while testing. Need a more surefire way of getting this to work

Rhials committed:
adjustments and dynamic setup

Rhials committed:
gives a free armblade to ease the pain

Rhials committed:
Merge remote-tracking branch 'upstream/master' into latejoin-ling

Rhials committed:
removes things i dont like

Rhials committed:
oops

Profakos commented:
We have a similar one to this on our downstream (the difference is that ours flies in via an Abandoned Pod and has tools and a chameleon jumpsuit) to go inside, and we have found that it works extremely well as a midround, even if it is really hard mode indeed. (Though we have different objectives where they have to suck dry people, so ours a bit more aggressive.)

SuperSlayer0 commented:
I suggest putting an image of a changeling in an organic spacesuit and an armblade rather then an engineer as a preference image

epicgamer545 commented:
it should say "OUR mind stirs to life" instead of "your" to fit the pattern of changeling flavor text

the-orange-cow commented:
Are the free fleshsuit and armblade permanent abilities, or refundable for extra points?

Rhials committed:
adds seperate space ling antag datum and mesage

Rhials committed:
yeah bitch yeab hitch

Fikou commented:
venom

blackdav123 commented:
add lings as latejoin antags too it makes no sense for traitor and heretic to have it and not ling

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -844,3 +844,24 @@
+/// Space Changeling ruleset
+/datum/dynamic_ruleset/midround/from_ghosts/changeling_midround
+	name = "Space Changeling"
+	midround_ruleset_style = MIDROUND_RULESET_STYLE_LIGHT
```
We're definitely going to have to keep a very close eye on this because changelings can be wicked deadly

Mothblocks commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -363,7 +363,7 @@
  * [sting_path] - the power that's being purchased / evolved.
  */
-/datum/antagonist/changeling/proc/purchase_power(datum/action/changeling/sting_path)
+/datum/antagonist/changeling/proc/purchase_power(datum/action/changeling/sting_path, free = FALSE)
```
I don't like boolean parameters like this. Instead, Make a new proc named `give_power` that runs all the code, then returns TRUE/FALSE if it succeeded (like this already does now).

Then, have `purchase_power` do something like:

```dm
var/success = give_power(sting_path)
if (success)
    genetic_points -= initial(sting_path.dna_cost)
return success
```

Mothblocks commented on code/modules/events/ghost_role/changeling_event.dm:
```diff
@@ -0,0 +1,90 @@
+	player_mind.transfer_to(new_changeling)
+	player_mind.special_role = ROLE_CHANGELING_MIDROUND
+	player_mind.add_antag_datum(/datum/antagonist/changeling/space)
+	SEND_SOUND(new_changeling, sound('sound/magic/mutate.ogg'))
```
```suggestion
	SEND_SOUND(new_changeling, 'sound/magic/mutate.ogg')
```

`sound` is only needed for dynamic names.

Mothblocks commented on code/modules/events/ghost_role/changeling_event.dm:
```diff
@@ -0,0 +1,90 @@
+	var/turf/picked_start = spaceDebrisStartLoc(start_side, start_z)
+
+	var/obj/effect/meteor/meaty/changeling/changeling_meteor = new/obj/effect/meteor/meaty/changeling(picked_start, get_random_station_turf())
+	var/mob/living/carbon/human/new_changeling = new /mob/living/carbon/human/(picked_start)
```
```suggestion
	var/mob/living/carbon/human/new_changeling = new(picked_start)
```

Mothblocks commented on code/modules/events/ghost_role/changeling_event.dm:
```diff
@@ -0,0 +1,90 @@
+	var/start_z = pick(SSmapping.levels_by_trait(ZTRAIT_STATION))
+	var/turf/picked_start = spaceDebrisStartLoc(start_side, start_z)
+
+	var/obj/effect/meteor/meaty/changeling/changeling_meteor = new/obj/effect/meteor/meaty/changeling(picked_start, get_random_station_turf())
```
```suggestion
	var/obj/effect/meteor/meaty/changeling/changeling_meteor = new(picked_start, get_random_station_turf())
```

Mothblocks commented on code/modules/events/ghost_role/changeling_event.dm:
```diff
@@ -0,0 +1,90 @@
+
+/proc/generate_changeling_meteor(mob/candidate)
+	var/mob/dead/selected = make_body(candidate) //Give the selected player a body, and grab their mind
+	var/datum/mind/player_mind = new /datum/mind(selected.key)
```
```suggestion
	var/datum/mind/player_mind = new(selected.key)
```

Mothblocks commented on code/modules/events/ghost_role/changeling_event.dm:
```diff
@@ -0,0 +1,90 @@
+	message_admins("[ADMIN_LOOKUPFLW(new_changeling)] has been made into a space changeling by an event.")
+	new_changeling.log_message("was spawned as a midround space changeling by an event.", LOG_GAME)
+
+	var/datum/antagonist/changeling/changeling_datum = locate(/datum/antagonist/changeling) in player_mind.antag_datums
```
```suggestion
	var/datum/antagonist/changeling/changeling_datum = locate() in player_mind.antag_datums
```

Mothblocks commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -1010,9 +1011,25 @@
+	name = "\improper Space Changeling"
+
+/datum/antagonist/changeling/space/get_preview_icon()
+	var/icon/final_icon = render_preview_outfit(/datum/outfit/changeling_space)
```
Can we at least give them an armblade or something, this looks so stupid

Ekaterina-von-Russland commented:
Very based. I mistook this for latejoin at first, which should also be added.

ArthuroPep commented:
> add lings as latejoin antags too it makes no sense for traitor and heretic to have it and not ling

I think this is the sole reason why i only got changeling once in 150 hours of play

Rhials committed:
Merge remote-tracking branch 'upstream/master' into latejoin-ling

Rhials committed:
lets get this over with

space lings start with the armblade out, because it makes the icon look less goofy and also makes sense already

Rhials committed:
removes a bunch of uncessesary bonus typecasting

also changes the free power logic to be less poo poo

Rhials committed:
screenie updatye

CCC23 commented:
Time to finally be able to play ling for once

san7890 commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -392,17 +392,31 @@
 		return FALSE
 
-	var/datum/action/changeling/new_action = new sting_path()
+	var/success = give_power(sting_path)
+	if(success)
+		genetic_points -= initial(sting_path.dna_cost)
+	return success
```
The `give_power()` proc that you introduce in this PR doesn't return TRUE at all, so this will _always_ be null, making line 397 unreachable. It will also mean that we always return a null value (falsey), which isn't nice to any consumers of this proc.

On second read, you document that give_power returns TRUE, but you appear to have forgotten to actually code that in. Please do fix this.

san7890 commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -392,17 +392,31 @@
-	purchased_powers[sting_path] = new_action
+	purchased_powers[power_path] = new_action
 	new_action.on_purchase(owner.current) // Grant() is ran in this proc, see changeling_powers.dm.
 	log_changeling_power("[key_name(owner)] adapted the [new_action] power")
```
```suggestion
	log_changeling_power("[key_name(owner)] adapted the [new_action] power")
	
	return TRUE
```

wraith-54321 commented:
How do Meteor Defense Systems effect this?

Rhials committed:
how the fuck did I miss that

Helg2 commented:
How will ling breath in space though?

SuperSlayer0 commented:
![image](https://user-images.githubusercontent.com/91609255/216833598-1e221caf-a060-441c-8903-1b426ee0f473.png)

Ekaterina-von-Russland commented:
It was us. World's first meteor lings, and we fluked so hard.

Ekaterina-von-Russland commented:
> How will ling breath in space though?

Salbutamol, apparently!
